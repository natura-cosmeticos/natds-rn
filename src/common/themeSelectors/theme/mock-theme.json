{
  "avatarSizes": {
    "tiny": {
      "size": 24,
      "fontSize": ".625rem"
    },
    "small": {
      "size": 32,
      "fontSize": ".875rem"
    },
    "standard": {
      "size": 40,
      "fontSize": "1rem"
    },
    "large": {
      "size": 48,
      "fontSize": "1rem"
    },
    "huge": {
      "size": 80,
      "fontSize": "1.5rem"
    }
  },
  "buttonSizes": {
    "small": {
      "paddingTop": 8,
      "paddingRight": 8,
      "paddingBottom": 8,
      "paddingLeft": 8,
      "minHeight": 32
    },
    "medium": {
      "paddingTop": 12,
      "paddingRight": 16,
      "paddingBottom": 12,
      "paddingLeft": 16,
      "minHeight": 40
    },
    "large": {
      "paddingTop": 16,
      "paddingRight": 16,
      "paddingBottom": 16,
      "paddingLeft": 16,
      "minHeight": 48
    }
  },
  "fontSize": {
    "h1": {
      "fontSize": "6rem",
      "fontWeight": 300
    },
    "h2": {
      "fontSize": "3.75rem",
      "fontWeight": 300
    },
    "h3": {
      "fontSize": "3rem",
      "fontWeight": 400
    },
    "h4": {
      "fontSize": "2.125rem",
      "fontWeight": 400
    },
    "h5": {
      "fontSize": "1.5rem",
      "fontWeight": 400
    },
    "h6": {
      "fontSize": "1.25rem",
      "fontWeight": 600
    },
    "subtitle1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "subtitle2": {
      "fontSize": "0.875rem",
      "fontWeight": 500
    },
    "body1": {
      "fontSize": "1rem",
      "fontWeight": 400
    },
    "body2": {
      "fontSize": "0.875rem",
      "fontWeight": 400
    },
    "button": {
      "fontSize": "0.875rem",
      "fontWeight": 600
    },
    "caption": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    },
    "overline": {
      "fontSize": "0.75rem",
      "fontWeight": 400
    }
  },
  "iconSizes": {
    "micro": 16,
    "tiny": 24,
    "small": 32,
    "standard": 48,
    "medium": 64,
    "large": 72,
    "largex": 96,
    "largexx": 128,
    "huge": 144,
    "hugex": 192,
    "hugexx": 256
  },
  "opacity": {
    "12": 0.12,
    "16": 0.16,
    "24": 0.24,
    "32": 0.32,
    "48": 0.48,
    "56": 0.56,
    "64": 0.64,
    "80": 0.8,
    "00": 0,
    "04": 0.04,
    "08": 0.08
  },
  "overlay": {
    "solidBlack": 0.56
  },
  "radius": {
    "none": 0,
    "circle": "50%",
    "small": 2,
    "medium": 4,
    "large": 8
  },
  "sizes": {
    "none": 0,
    "micro": 4,
    "tiny": 8,
    "small": 16,
    "standard": 24,
    "semi": 32,
    "semix": 40,
    "medium": 48,
    "mediumx": 56,
    "large": 64,
    "largex": 72,
    "largexx": 80,
    "largexxx": 88,
    "huge": 96,
    "hugex": 128,
    "hugexx": 144,
    "hugexxx": 192,
    "veryhuge": 256
  },
  "spacing": {
    "spacing": 8,
    "spacingNone": 0,
    "spacingMicro": 4,
    "spacingTiny": 8,
    "spacingSmall": 16,
    "spacingStandard": 24,
    "spacingSemi": 32,
    "spacingLarge": 48,
    "spacingHuge": 64
  },
  "colorTokens": {
    "colorPrimary": "#F4AB34",
    "colorOnPrimary": "#333333",
    "colorPrimaryLight": "#FEFDE8",
    "colorOnPrimaryLight": "#333333",
    "colorPrimaryDark": "#EF8426",
    "colorOnPrimaryDark": "#333333",
    "colorSecondary": "#FF6B0B",
    "colorOnSecondary": "#000000",
    "colorSecondaryLight": "#FFF8E1",
    "colorOnSecondaryLight": "#000000",
    "colorSecondaryDark": "#FF5808",
    "colorOnSecondaryDark": "#000000",
    "colorBackground": "#FAFAFA",
    "colorOnBackground": "#333333",
    "colorSurface": "#FFFFFF",
    "colorOnSurface": "#333333",
    "colorHighlight": "#000000",
    "colorHighEmphasis": "#333333",
    "colorMediumEmphasis": "#777777",
    "colorLowEmphasis": "#BBBBBB",
    "colorLink": "#227BBD",
    "colorOnLink": "#FFFFFF",
    "colorSuccess": "#569A32",
    "colorOnSuccess": "#FFFFFF",
    "colorWarning": "#FCC433",
    "colorOnWarning": "#333333",
    "colorAlert": "#E74627",
    "colorOnAlert": "#FFFFFF"
  },
  "elevation": {
    "1": {
      "elevation": 1,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 1, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 1
    },
    "2": {
      "elevation": 2,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 2, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 2
    },
    "3": {
      "elevation": 3,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 3, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 4
    },
    "4": {
      "elevation": 4,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 4, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 5
    },
    "5": {
      "elevation": 5,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 6, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 10
    },
    "6": {
      "elevation": 6,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 8, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 10
    },
    "7": {
      "elevation": 7,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 9, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 12
    },
    "8": {
      "elevation": 8,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 12, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 17
    },
    "9": {
      "elevation": 9,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 16, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 24
    },
    "10": {
      "elevation": 10,
      "shadowColor": "#000000",
      "shadowOffset": { "height": 24, "width": 0 },
      "shadowOpacity": 0.14,
      "shadowRadius": 38
    }
  }
}
