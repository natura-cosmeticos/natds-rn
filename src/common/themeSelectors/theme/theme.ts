import { tokens as styleTokens } from '@naturacosmeticos/natds-styles';
import { IColorThemeTokens } from '@naturacosmeticos/natds-styles/dist/tokens/colors/themeBound/IColorThemeTokens';
import { IElevation } from '@naturacosmeticos/natds-styles/dist/tokens/elevation/IElevation';

type Tokens = typeof styleTokens;
type TokensWithoutColor = Omit<Tokens, 'colors'>
type TokensWithoutColorTokens = Omit<TokensWithoutColor, 'colorTokens'>
type TokensWithoutElevation = Omit<TokensWithoutColorTokens, 'elevation'>

interface ElevationProps {
  elevation: number
  shadowColor: string
  shadowOffset: { height: number, width: number }
  shadowOpacity: number
  shadowRadius: number
}

type Elevation = Record<keyof IElevation, ElevationProps>


export type Theme = TokensWithoutElevation
  & {
    colorTokens: IColorThemeTokens,
    elevation: Elevation | {}
  }

type ThemeMode = 'dark' | 'light';

const parseElevation = (elevation: string) => {
  const regEx = /(\d+(\.\d+)?)/g;
  const matches = elevation.match(regEx);

  if (matches) {
    const [width, height, blurRadius, , r, g, b, alpha] = matches;

    const toUnit = (prop: string) => parseInt(prop, 0);

    return {
      elevation: 2,
      shadowColor: `rgb(${r}, ${g}, ${b})`,
      shadowOffset: { height: toUnit(height), width: toUnit(width) },
      shadowOpacity: parseFloat(alpha),
      shadowRadius: toUnit(blurRadius),
    };
  }

  return {};
};

const buidElevation = (elevation: IElevation) => {
  const nativeElevation = {};

  Object
    .keys(elevation)
    .forEach((key) => { nativeElevation[key] = parseElevation(elevation[key]); });

  return nativeElevation;
};

export const buildTheme = (tokens: Tokens, brand: string, mode: ThemeMode) => {
  const colorThemeName = `${brand}${mode === 'light' ? 'Light' : 'Dark'}ColorTokens`;
  const localTokens = { ...tokens };

  delete localTokens.colors;

  const theme: Theme = {
    ...localTokens,
    colorTokens: localTokens.colorTokens[colorThemeName],
    elevation: buidElevation(localTokens.elevation),
  };

  return theme;
};

interface ThemeSelector {
  (theme: Theme): string | undefined
}

export const getTheme = (selector: ThemeSelector) => (
  { theme }: { theme: Theme },
) => selector(theme);
