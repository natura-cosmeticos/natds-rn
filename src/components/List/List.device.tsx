import { storiesOf } from '@storybook/react-native';
import {
  Base,
  Dividers,
  RippleFeedback,
  SelectionFeedback,
} from './List.stories';

storiesOf('List', module)
  .add('Base', Base)
  .add('FeedbackRipple', RippleFeedback)
  .add('FeedbackSelection', SelectionFeedback)
  .add('Dividers', Dividers);
