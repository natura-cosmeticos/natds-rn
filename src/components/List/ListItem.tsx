/* eslint-disable max-lines */
import React from 'react';
import { View } from 'react-native';
import { withTheme } from 'styled-components';

import { Theme } from '../../common/themeSelectors';
import { Icon } from '../Icon';

import {
  HeaderWrapper,
  HeaderTitle,
  IconPress,
  TouchableOpacity,
  HeaderTitleView,
  dividerContainerStyles,
  Divider,
} from './ListItem.styles';

export interface ListItemProps {
  /**
   * An active item is usable
   */
  active: boolean;
  /**
   * A disabled item is unusable
   */
  disabled?: boolean;
  /**
   * Show a separator between the items
   */
  divider: boolean;
  /**
   * Divider's type options
   */
  dividerType?: 'full-bleed' | 'inset' | 'middle';
  /**
   * Wether show or not the left icon
   */
  hideIconLeft: boolean;
  /**
   * Wether show or not the right icon
   */
  hideIconRight: boolean;
  /**
   * Left icon name
   */
  iconLeft?: string;
  /**
   * Right icon name
   */
  iconRight?: string;
  /**
   * onPress event handler
   */
  onPress?(): void;
  /**
   * Handle click on the left icon
   */
  onPressLeft?(): void;
  /**
   * Handle click on the right icon
   */
  onPressRight?(): void;
  /**
   * Id for testing
   */
  testID?: string;
  /**
   * App's theme
   */
  theme: Theme;
  /**
   * Item's title
   */
  title: string;
}

const ListItemComponent = ({
  title,
  disabled,
  iconLeft,
  iconRight,
  hideIconLeft,
  hideIconRight,
  onPressRight,
  onPressLeft,
  onPress,
  divider,
  dividerType,
  active,
  theme,
  testID = 'listId',
}: ListItemProps) => {
  if (!title) {
    throw new Error('Title should not be an empty string');
  }

  const renderIcon = (iconData, iconPress, position, disabledIcon) => {
    let styleCustom = {};

    if (position === 'right') {
      styleCustom = { position: 'absolute', right: 0 };
    } else {
      styleCustom = { left: 0, position: 'absolute' };
    }

    return (
      <View style={styleCustom}>
        <IconPress
          testID={`${testID}-icon-${position}`}
          onPress={iconPress}
          disabled={disabledIcon}
          position={position}>
          <Icon
            color={disabledIcon ? 'lowEmphasis' : 'default'}
            size="standard"
            name={iconData}
          />
        </IconPress>
      </View>
    );
  };

  const localPress = () => {
    if (onPress) {
      onPress();
    }
  };

  return (
    <View>
      <HeaderWrapper
        style={dividerContainerStyles({ dividerType, theme })}
        accessible
        accessibilityRole="header"
        testID={testID}
        active={active}>
        {!hideIconLeft && renderIcon(iconLeft, onPressLeft, 'left', disabled)}

        <HeaderTitleView
          hideIconLeft={hideIconLeft}
          hideIconRight={hideIconRight}>
          <TouchableOpacity
            disabled={disabled}
            onPress={localPress}
            testID={`${testID}-title`}>
            <HeaderTitle disabled={disabled} accessibilityLabel={`${title}`}>
              {title}
            </HeaderTitle>
          </TouchableOpacity>
        </HeaderTitleView>

        {!hideIconRight && renderIcon(iconRight, onPressRight, 'right', disabled)}
      </HeaderWrapper>

      {divider && (
        <View style={dividerContainerStyles({ dividerType, theme })}>
          <Divider />
        </View>
      )}
    </View>
  );
};

export const ListItem = withTheme(ListItemComponent);
