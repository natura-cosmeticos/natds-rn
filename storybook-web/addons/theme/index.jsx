import React, { useEffect, Fragment } from 'react';
import { useAddonState, Consumer } from '@storybook/api';
import { TabButton, WithTooltip, TooltipLinkList } from '@storybook/components';
import {
  PANEL_ID, CHANGE, PARAM_KEY, getStoryBookTheme,
} from './shared';
import { light, dark } from '../../theme';

import './styles.css';

const DEFAULT_THEME = {
  name: 'natura',
  type: 'light',
};

const buildThemeProps = mode => ({ ...DEFAULT_THEME, type: mode });

function parseTheme(themes) {
  return Object.entries(themes).reduce((accum, [name, variants]) => {
    Object.entries(variants).forEach(([type]) => {
      accum.push({ name, type });
    });

    return accum;
  }, []);
}

const createBackgroundSelectorItem = ({ name, type }, onChange, selectedItem) => ({
  active: selectedItem.name === name && selectedItem.type === type,
  id: `${name} - ${type}`,
  onClick: () => onChange({ name, type }),
  title: `${name} - ${type}`,
});

const getDisplayedItems = (list, onChange, selectedItem) => {
  if (!list.length) {
    return [];
  }

  return [
    ...list.map(params => createBackgroundSelectorItem(params, onChange, selectedItem)),
  ];
};

const mapper = ({ api, state }) => {
  const story = state.storiesHash[state.storyId];
  const { themes = {} } = (story && api.getParameters(story.id, PARAM_KEY)) || {};

  return { items: parseTheme(themes) };
};

export default function Theme({ channel, api }) {
  const apiTheme = getStoryBookTheme();
  const initialTheme = apiTheme ? buildThemeProps(apiTheme) : DEFAULT_THEME;
  const [currentTheme, changeTheme] = useAddonState(PANEL_ID, initialTheme);
  const { disabled } = api.getCurrentParameter(PARAM_KEY) || {};

  const handleChange = (params) => {
    changeTheme(params);
    api.setOptions({
      theme: params.type === 'light' ? light : dark,
    });
    const event = new Event('CHANGE_THEME', { detail: params.type });

    window.dispatchEvent(event);
    channel.emit(CHANGE, params);
  };

  const renderTooltip = items => ({ onHide }) => (
    <TooltipLinkList
      links={getDisplayedItems(items, (index) => {
        handleChange(index);
        onHide();
      }, currentTheme)}
    />
  );

  const eventListener = event => handleChange(buildThemeProps(event.matches ? 'dark' : 'light'));

  useEffect(() => {
    handleChange(initialTheme);

    const matchMediaEnvent = window.matchMedia('(prefers-color-scheme: dark)');

    matchMediaEnvent.addEventListener('change', eventListener);

    return () => {
      matchMediaEnvent.removeEventListener('change', eventListener);
    };
  }, []);

  if (disabled) return null;

  return (
    <Consumer filter={mapper}>
      {({ items }) => (items.length ? (
          <Fragment>
            <WithTooltip
              placement="top"
              trigger="click"
              closeOnClick
              tooltip={renderTooltip(items)}
            >
              <TabButton
                key="Theme"
                title="Change the Theme">
                {currentTheme.name} - {currentTheme.type}
              </TabButton>
            </WithTooltip>
          </Fragment>
      ) : null)}
    </Consumer>
  );
}
