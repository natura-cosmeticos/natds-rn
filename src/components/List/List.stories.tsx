/* eslint-disable max-lines */
import React, { useState } from 'react';
import { View, FlatList, Text } from 'react-native';
import { Card } from '../Card';
import { Divider } from '../Divider';
import { ListItem, ListItemFeedback } from './ListItem';
import { Checkbox } from '../Checkbox';
import { TextWithTheme } from '../../common/HelperComponents/ThemeHelper.styles';
import { data } from './stories.data';

export default {
  component: ListItem,
  parameters: {
    componentSubtitle: 'List let users choose between several items',
  },
  title: 'Components|List',
};

export const Base = () => (
  <FlatList
    data={data}
    renderItem={({ item }) => (
      <ListItem>
        <View style={{ padding: 16 }}>
          <TextWithTheme>{item.title}</TextWithTheme>
        </View>
      </ListItem>
    )}
  />
);

export const RippleFeedback = () => (
  <FlatList
    data={data}
    renderItem={({ item }) => (
      <ListItem feedback="ripple" onPress={() => console.log('I was pressed')}>
        <View style={{ padding: 16 }}>
          <TextWithTheme>{item.title}</TextWithTheme>
        </View>
      </ListItem>
    )}
  />
);

export const SelectionFeedback = () => {
  const [selected, setSelected] = useState('');

  const renderItem = ({ item }) => {
    const listItemProps = {
      feedback: 'selection' as ListItemFeedback,
      selected: item.key === selected,
      ...(item.unreleased ? {} : { onPress: () => setSelected(item.key) }),
    };

    return (
      <ListItem {...listItemProps}>
        <View style={{
          flexDirection: 'row',
          justifyContent: 'space-between',
          padding: 16,
        }}>
          <TextWithTheme>{item.title}</TextWithTheme>
          <Checkbox
            selected={item.key === selected}
            onPress={() => setSelected(item.key)}
            disabled={item.unreleased}
          />
        </View>
      </ListItem>
    );
  };

  return (
    <FlatList
      ItemSeparatorComponent={Divider}
      data={data}
      renderItem={renderItem}
      extraData={selected}
    />
  );
};


export const Dividers = () => {
  const renderExtraInfo = (extraInfo) => {
    if (!extraInfo) return null;

    const extraInfoData = Object
      .keys(extraInfo)
      .map((item, index) => (
        <View key={extraInfo[item].key} style={{ backgroundColor: '#fafafa' }}>
          <ListItem>
            <View style={{
              flexDirection: 'row',
              justifyContent: 'space-between',
              paddingHorizontal: 32,
              paddingVertical: 16,
            }}>
              <TextWithTheme>{item}</TextWithTheme>
              <TextWithTheme>{extraInfo[item]}</TextWithTheme>
            </View>
          </ListItem>
          {
            index < Object.keys(extraInfo).length - 1
              ? <Divider type="inset" />
              : null
          }
        </View >
      ));

    return extraInfoData;
  };

  return (
    <View style={{
      flex: 1,
      paddingHorizontal: 10,
      paddingVertical: 20,
    }}>
      <Card>
        <FlatList
          ItemSeparatorComponent={Divider}
          data={data}
          renderItem={({ item }) => (
            <>
              <ListItem onPress={() => console.log('I was pressed')}>
                <View style={{ padding: 16 }}>
                  <Text>{item.title}</Text>
                </View>
              </ListItem>
              {renderExtraInfo(item.extraInfo)}
            </>
          )}
        />
      </Card>
    </View>
  );
};
