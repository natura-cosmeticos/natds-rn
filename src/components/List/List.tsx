/* eslint-disable max-lines */
import React, { useState } from 'react';
import { ScrollView } from 'react-native';
import { withTheme } from 'styled-components';
import { Theme } from '../../common/themeSelectors';

import { ListItem } from './ListItem';

export interface ItemProps {
  /**
   * A disabled item is unusable
   */
  disabled?: boolean;
  /**
   * Show a separator between the items
   */
  divider: boolean;
  /**
   * Divider's type options
   */
  dividerType?: 'fullBleed' | 'inset' | 'middle';
  /**
   * Wether show or not the left icon
   */
  hideIconLeft: boolean;
  /**
   * Wether show or not the right icon
   */
  hideIconRight: boolean;
  /**
   * Left icon name
   */
  iconLeft?: string;
  /**
   * Right icon name
   */
  iconRight?: string;
  /**
   * Item's id
   */
  id: string;
  /**
   * onPress event handler
   */
  onPress?(): void;
  /**
   * Handle click on the left icon
   */
  onPressLeft?(): void;
  /**
   * Handle click on the right icon
   */
  onPressRight?(): void;
  /**
   * Item's title
   */
  title: string;
}

export interface ListProps {
  /**
   * A disabled item is unusable
   */
  disabled?: boolean;
  /**
   * Wether the list is required to have a selecte item
   */
  isRequired?: boolean;
  /**
   * List's items
   */
  items: Array<ItemProps>;
  /**
   * Wether you can select more than one item from the list
   */
  multiSelect?: boolean;
  /**
   * A function to handle when the selected item is changed
   */
  onChange(): void;
  /**
   * Selected item's name
   */
  selected: Array<string>;
  /**
   * Id for testing
   */
  testID?: string;
  /**
   * App's theme
   */
  theme: Theme;
}

const ListComponent = ({
  disabled,
  isRequired,
  items,
  multiSelect,
  onChange,
  selected,
  testID = 'listId',
  theme,
}: ListProps) => {
  const [currentSelected, setCurrentSelected] = useState(selected);

  const updateCurrentSelectedMultiple = (item) => {
    let result = currentSelected;

    if (currentSelected.includes(item.id)) {
      // remove from selection
      result = currentSelected.filter(select => select !== item.id);
    } else {
      // add item to list
      result = [...currentSelected, ...[item.id]];
    }

    setCurrentSelected(result);

    return result;
  };

  const updateCurrentSelectedSingle = (item) => {
    let result = currentSelected;

    if (currentSelected.includes(item.id)) {
      // remove from selection only list is not required
      if (!isRequired) {
        result = currentSelected.filter(select => select !== item.id);
      }
    } else {
      // add item to list
      result = [...[item.id]];
    }

    setCurrentSelected(result);

    return result;
  };

  const onPress = (callback, item) => {
    let selectedItems: Array<string> = [];

    if (multiSelect) {
      selectedItems = updateCurrentSelectedMultiple(item);
    } else {
      selectedItems = updateCurrentSelectedSingle(item);
    }

    if (callback) {
      callback(selectedItems);
    }

    if (item.onPress) {
      item.onPress(item);
    }
  };

  const renderItem = item => (
    <ListItem
      disabled={disabled}
      title={item.title}
      iconLeft={item.iconLeft}
      iconRight={item.iconRight}
      hideIconLeft={item.hideIconLeft}
      hideIconRight={item.hideIconRight}
      onPress={() => onPress(onChange, item)}
      onPressLeft={item.onPressLeft}
      onPressRight={item.onPressRight}
      divider={item.divider}
      dividerType={item.dividerType}
      active={currentSelected.includes(item.id)}
      key={item.id}
      theme={theme}
      testID={testID}
    />
  );

  return (
    <ScrollView keyboardShouldPersistTaps="always">
      {items.map(item => renderItem(item))}
    </ScrollView>
  );
};

export const List = withTheme(ListComponent);
