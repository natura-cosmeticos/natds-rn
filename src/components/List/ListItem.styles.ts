import styled from 'styled-components/native';
import {
  getTheme,
  getSpacingSmall,
  getColorPrimary,
  getOpacityMediumLow,
} from '../../common/themeSelectors';
import { ListItemProps } from './ListItem';

const inlineStyles = ({ inline }: ListItemProps) => (
  inline
    ? 'flex-direction: row; justify-content: space-between;'
    : ''
);

const selectedBackground = ({ selected, theme }: Required<ListItemProps>) => (
  selected
    ? getColorPrimary(theme) + getOpacityMediumLow(theme) * 100
    : 'transparent'
);


export const ListItem = styled.View<ListItemProps>`
  background: ${selectedBackground};
  ${inlineStyles}
`;
