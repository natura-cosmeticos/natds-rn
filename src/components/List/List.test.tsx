import '@testing-library/jest-native/extend-expect';
import React from 'react';
import { render } from '@testing-library/react-native';
import { ThemeProvider } from 'styled-components/native';
import theme from '../../common/themeSelectors/theme/mock-theme.json';

import { List, ListProps, ItemProps } from './List';

jest.mock('../../common/themeSelectors', () => ({
  getColorBackground: () => '#FFFFFF',
  getColorHighEmphasis: () => '#FAF3E3',
  getColorLowEmphasis: () => '#FEEEEF',
  getColorPrimary: () => '#FFFFFF',
  getSpacingTiny: () => 10,
}));

const renderList = (fn, props: ListProps) => (
  fn(
    <ThemeProvider theme={theme}>
      <List {...props} />
    </ThemeProvider>,
  )
);

describe('List component', () => {
  const listItems: ItemProps[] = [
    {
      divider: true,
      hideIconLeft: true,
      hideIconRight: true,
      id: '1',
      onPress: () => {},
      title: 'Title Text',
    },
    {
      divider: true,
      hideIconLeft: true,
      hideIconRight: true,
      id: '2',
      onPress: () => {},
      title: 'Title Text',
    },
    {
      divider: true,
      hideIconLeft: true,
      hideIconRight: true,
      id: '3',
      onPress: () => {},
      title: 'Title Text',
    },
    {
      divider: true,
      hideIconLeft: true,
      hideIconRight: true,
      id: '4',
      onPress: () => {},
      title: 'Title Text',
    },
  ];

  it('Should render List component correctly', () => {
    const defaultProps: ListProps = {
      isRequired: false,
      items: listItems,
      multiSelect: false,
      onChange: () => {},
      selected: [listItems[0].id],
      theme,
    };

    const list = renderList(render, defaultProps).asJSON();

    expect(list).toMatchSnapshot('List snapshot');
  });
});
