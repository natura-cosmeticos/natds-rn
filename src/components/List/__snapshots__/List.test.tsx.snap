// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component Should render List component correctly: List snapshot 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <ScrollView
    keyboardShouldPersistTaps="always"
  >
    <View>
      <View
        accessibilityRole="header"
        accessible={true}
        active={true}
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF29",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
            },
          ]
        }
        testID="listId"
      >
        <View
          hideIconLeft={true}
          hideIconRight={true}
          style={
            Array [
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "right": 0,
                "top": 0,
              },
            ]
          }
        >
          <
            style={
              Array [
                Object {
                  "alignSelf": "stretch",
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "justifyContent": "center",
                },
              ]
            }
            testID="listId-title"
          >
            <Text
              accessibilityLabel="Title Text"
              style={
                Array [
                  Object {
                    "color": "#FAF3E3",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "fontSize": 16,
                    "justifyContent": "center",
                    "marginLeft": 0,
                    "paddingBottom": 12,
                    "paddingLeft": 16,
                    "paddingTop": 12,
                  },
                ]
              }
            >
              Title Text
            </Text>
          </>
        </View>
      </View>
      <View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#FEEEEF",
                "height": 1,
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
    <View>
      <View
        accessibilityRole="header"
        accessible={true}
        active={false}
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF29",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
            },
          ]
        }
        testID="listId"
      >
        <View
          hideIconLeft={true}
          hideIconRight={true}
          style={
            Array [
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "right": 0,
                "top": 0,
              },
            ]
          }
        >
          <
            style={
              Array [
                Object {
                  "alignSelf": "stretch",
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "justifyContent": "center",
                },
              ]
            }
            testID="listId-title"
          >
            <Text
              accessibilityLabel="Title Text"
              style={
                Array [
                  Object {
                    "color": "#FAF3E3",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "fontSize": 16,
                    "justifyContent": "center",
                    "marginLeft": 0,
                    "paddingBottom": 12,
                    "paddingLeft": 16,
                    "paddingTop": 12,
                  },
                ]
              }
            >
              Title Text
            </Text>
          </>
        </View>
      </View>
      <View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#FEEEEF",
                "height": 1,
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
    <View>
      <View
        accessibilityRole="header"
        accessible={true}
        active={false}
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF29",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
            },
          ]
        }
        testID="listId"
      >
        <View
          hideIconLeft={true}
          hideIconRight={true}
          style={
            Array [
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "right": 0,
                "top": 0,
              },
            ]
          }
        >
          <
            style={
              Array [
                Object {
                  "alignSelf": "stretch",
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "justifyContent": "center",
                },
              ]
            }
            testID="listId-title"
          >
            <Text
              accessibilityLabel="Title Text"
              style={
                Array [
                  Object {
                    "color": "#FAF3E3",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "fontSize": 16,
                    "justifyContent": "center",
                    "marginLeft": 0,
                    "paddingBottom": 12,
                    "paddingLeft": 16,
                    "paddingTop": 12,
                  },
                ]
              }
            >
              Title Text
            </Text>
          </>
        </View>
      </View>
      <View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#FEEEEF",
                "height": 1,
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
    <View>
      <View
        accessibilityRole="header"
        accessible={true}
        active={false}
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF29",
              "flexBasis": 0,
              "flexGrow": 1,
              "flexShrink": 1,
              "paddingBottom": 0,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 0,
            },
          ]
        }
        testID="listId"
      >
        <View
          hideIconLeft={true}
          hideIconRight={true}
          style={
            Array [
              Object {
                "alignItems": "center",
                "bottom": 0,
                "justifyContent": "center",
                "left": 0,
                "right": 0,
                "top": 0,
              },
            ]
          }
        >
          <
            style={
              Array [
                Object {
                  "alignSelf": "stretch",
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "justifyContent": "center",
                },
              ]
            }
            testID="listId-title"
          >
            <Text
              accessibilityLabel="Title Text"
              style={
                Array [
                  Object {
                    "color": "#FAF3E3",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "fontSize": 16,
                    "justifyContent": "center",
                    "marginLeft": 0,
                    "paddingBottom": 12,
                    "paddingLeft": 16,
                    "paddingTop": 12,
                  },
                ]
              }
            >
              Title Text
            </Text>
          </>
        </View>
      </View>
      <View>
        <View
          style={
            Array [
              Object {
                "backgroundColor": "#FEEEEF",
                "height": 1,
                "width": "100%",
              },
            ]
          }
        />
      </View>
    </View>
  </ScrollView>
</View>
`;
